### ISSUE TITLE
aws-sigv4 depends unnecessarily upon aws-eventstream

### ISSUE THREAD

**Issue by hatch-carl** [aws/aws-sdk-ruby#2003]
Regarding #1946 

Respectfully, I do not appreciate this change.

My system depends upon the `aws-sigv4` gem as our means of signing requests to the AWS api. I chose to use the gem directly instead of a larger SDK gem because the SDK gem introduced too much bloat and additional dependencies to for our case, which is a very narrow usage of the S3 API.

Until now I have enjoyed relying on `aws-sigv4` as a low-overhead and minimal way to ensure that our requests are properly signed, relieving me of the effort of developing a functional signer myself.

However, by introducing the `aws-eventstream` dependency in this signing gem you have now burdened me with that gem, whose functionality I will never use, so that this gem can support signing event streams for that gem. I now have installed a gem with three commits, less than a year old, coded for an unknown AWS service.

What parts of `aws-sigv4` were essential to the implementation of signing events? Merely the `get_credentials` and `secret_access_key` provided by that object. The only other reason is the concept that signing stuff should go in the signing gem. You could have put this whole diff in the `aws-eventstream` gem, provided the secret key, and not disturbed me or any other users of this gem with a new dependency.

Further, this doesn't return a `Signature`, like the `sign_request` method. Why not? Well, the `Signature` class has a HTTP-centric view of Signature Version 4. Is Signature Version 4 expanding beyond HTTP? Unknown, and not documented. [The Documentation refrences only HTTP](https://docs.aws.amazon.com/general/latest/gr/signing_aws_api_requests.html). That makes this addition even more foreign.

Let's presume that the signing stuff must go in the signing gem. How could this be implemented such that consumers of the gem don't get a gem they don't need, and `aws-eventstream` users can sign their events?

```ruby
def sign_event(prior_signature, encoded_headers, payload, time)
  datetime = time.utc.strftime('%Y%m%dT%H%M%SZ')
  date = datetime[0, 8]
  sts = event_string_to_sign(datetime, prior_signature, encoded_headers, payload)
  signature(get_credentials.secret_access_key, date, sts)
end

def event_string_to_sign(datetime, prior_signature, encoded_headers, payload)
  [
    'AWS4-HMAC-SHA256-PAYLOAD',
    datetime,
    credential_scope(datetime[0, 8]),
    prior_signature,
    sha256_hexdigest(encoded_headers),
    sha256_hexdigest(payload)
  ].join("\n")
end
```

The only thing not to like about this is that there is some added encumbrance in that the users of`aws-eventstream` have to manipulate their own headers. I have no problem with this encumbrance because I am not a user of `aws-eventstream`.  I also presume that users of such know very well how to manipulate payload headers, encode things, and satisfy undocumented signing methods.

In conclusion one might advise me not to update the gem. That's good advice, and I hope that I never need to update the gem to correct a bug in Signature Version 4, gain a performance benefit, or fix a security issue. In such cases I'll be implementing the Signature Version 4 myself, and not bothering to leave a long-winded complaint in the issues page.

**Comment by cjyclaire** [comment]
`aws-sigv4` gem is intended for signature V4 computation gem for all AWS supported protocols, newly introduced changes is for support AWS eventstream protocol over HTTP2. 
`aws-eventstream` gem is a utility gem only for encode and decode binary streams which shouldn’t be messed with signing part. 

In terms of usage of `aws-eventstream` gem, if a customer want to use this gem without SDK, eventheaders and eventpayload have to be taken care of by their own, we have no knowledge of that, there is no “default” headers pattern per each event (unlike http1.1).

**Comment by hatch-carl** [comment]
If the `aws-eventstream` gem shouldn't be concerned with signing, then I argue that the `aws-sigv4` shouldn't be concerned with encoding and decoding binary streams. The natural conclusion, therefore, is that the caller is responsible for providing the glue which binds encoding and decoding binary streams to signing. This glue is the construction of the encoded header, and the addition of the signature to the final header.

**Comment by awood45** [comment]
Thank you for your feedback. At this point, extracting the aws-eventstream gem would be a breaking change. On balance there are also benefits to a central source of truth for signing, including for signing in the case of streamed operations. One has to depend on the other, it's a tradeoff either way.

Did you have particular concerns about the documentation of particular behavior/functions we could address?

**Comment by hatch-carl** [comment]
Thank you for your consideration and time in this mattern. I understand that now that the gem has been published there is an indeterminate risk of some users depending upon the added functionality, and to yank it back or significantly change its API may cause further dissatisfaction.

By this point I believe you understand my position, which I think would be shared by many, that taking on a gem dependency in a project is not something we do lightly. To depend on open source software is to own it cooperatively with the community, and so to publish open source software too implies a measure of cooperation. I don't mean to say that you should feel hamstrung by the whims and preferences of the community, that would be awful. But do take care especially when making changes, especially changes that ask us, the consumers of the gem, to extend our trust into further dependencies.
